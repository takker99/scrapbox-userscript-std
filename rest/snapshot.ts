import type {
  ErrorLike,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  PageSnapshotList,
  PageSnapshotResult,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { type BaseOptions, setDefaults } from "./options.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import type { FetchError } from "./mod.ts";
import type { TargetedResponse } from "./targeted_response.ts";
import {
  createErrorResponse,
  type createSuccessResponse as _createSuccessResponse,
  createTargetedResponse,
} from "./utils.ts";

/** 不正な`timestampId`を渡されたときに発生するエラー */
export interface InvalidPageSnapshotIdError extends ErrorLike {
  name: "InvalidPageSnapshotIdError";
}

export type SnapshotError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | InvalidPageSnapshotIdError
  | HTTPError;

/** get a page snapshot
 *
 * @param options connect.sid etc.
 */
export const getSnapshot = async (
  project: string,
  pageId: string,
  timestampId: string,
  options?: BaseOptions,
): Promise<
  TargetedResponse<
    200 | 400 | 404 | 422,
    PageSnapshotResult | SnapshotError | FetchError
  >
> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}/${timestampId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  const response = createTargetedResponse<200 | 400 | 404 | 422, SnapshotError>(
    res,
  );

  if (response.status === 422) {
    return createErrorResponse(422, {
      name: "InvalidPageSnapshotIdError",
      message: await response.text(),
    });
  }

  await parseHTTPError(response, [
    "NotFoundError",
    "NotLoggedInError",
    "NotMemberError",
  ]);

  return response;
};

export type SnapshotTimestampIdsError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | HTTPError;

/**
 * Retrieves the timestamp IDs for a specific page in a project.
 *
 * @param project - The name of the project.
 * @param pageId - The ID of the page.
 * @param options - Optional configuration options.
 * @returns A promise that resolves to a {@link Result} object containing the page snapshot list if successful,
 * or an error if the request fails.
 */
export const getTimestampIds = async (
  project: string,
  pageId: string,
  options?: BaseOptions,
): Promise<
  TargetedResponse<
    200 | 400 | 404,
    PageSnapshotList | SnapshotTimestampIdsError | FetchError
  >
> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  const response = createTargetedResponse<
    200 | 400 | 404,
    SnapshotTimestampIdsError
  >(res);

  await parseHTTPError(response, [
    "NotFoundError",
    "NotLoggedInError",
    "NotMemberError",
  ]);

  return response;
};
