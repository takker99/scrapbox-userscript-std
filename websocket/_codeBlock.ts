import type { TinyCodeBlock } from "../rest/getCodeBlocks.ts";

/** Interface for storing code block title line information
 *
 * In Scrapbox, code blocks start with a title line that defines:
 * - The code's filename or language identifier
 * - Optional language specification in parentheses
 * - Indentation level for nested blocks
 */
export interface CodeTitle {
  filename: string;
  lang: string;
  indent: number;
}

/** Extract properties from a code block title line
 *
 * This function parses a line of text to determine if it's a valid code block title.
 * Valid formats include:
 * - `code:filename.ext` - Language determined by extension
 * - `code:filename(lang)` - Explicit language specification
 * - `code:lang` - Direct language specification without filename
 *
 * @param lineText - The line text to parse
 * @returns A {@linkcode CodeTitle} | {@linkcode null}:
 *          - Success: A {@linkcode CodeTitle} object containing filename and language info
 *          - Error: {@linkcode null} if the line is not a valid code block title
 *                           and indentation level.
 */
export const extractFromCodeTitle = (lineText: string): CodeTitle | null => {
  const matched = lineText.match(/^(\s*)code:(.+?)(\(.+\)){0,1}\s*$/);
  if (matched === null) return null;
  const filename = matched[2].trim();
  let lang = "";
  if (matched[3] === undefined) {
    const ext = filename.match(/.+\.(.*)$/);
    if (ext === null) {
      // `code:ext`
      lang = filename;
    } else if (ext[1] === "") {
      // Reject "code:foo." format as it's invalid (trailing dot without extension)
      // This ensures code blocks have either a valid extension or no extension at all
      return null;
    } else {
      // `code:foo.ext`
      lang = ext[1].trim();
    }
  } else {
    lang = matched[3].slice(1, -1);
  }
  return {
    filename: filename,
    lang: lang,
    indent: matched[1].length,
  };
};

/** Calculate the indentation level for code block content
 *
 * The content of a code block is indented one level deeper than its title line.
 * This function determines the correct indentation by analyzing the title line's
 * whitespace and adding one additional level.
 */
export function countBodyIndent(
  codeBlock: Pick<TinyCodeBlock, "titleLine">,
): number {
  return codeBlock.titleLine.text.length -
    codeBlock.titleLine.text.trimStart().length + 1;
}
