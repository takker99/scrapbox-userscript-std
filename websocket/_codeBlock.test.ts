import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { extractFromCodeTitle } from "./_codeBlock.ts";

/**
 * Tests for code block title parsing functionality
 *
 * These tests verify the parsing of code block titles in various formats:
 * - Valid formats: code:filename.ext(param), code:filename(param), code:filename.ext
 * - Invalid formats: trailing dots, incorrect prefixes, non-code blocks
 */
Deno.test("extractFromCodeTitle()", async (t) => {
  await t.step("accurate titles", async (st) => {
    const titles = [
      "code:foo.extA(extB)", // Basic format: no spaces
      " code:foo.extA(extB)", // Leading space before code:
      "  code: foo.extA (extB)", // Spaces around components
      "  code: foo (extB) ", // Extension omitted, has parameter
      "  code: foo.extA ", // Extension only, no parameter
      "  code: foo ", // Basic name only
      "  code: .foo ", // Leading dot in name
    ];
    for (const title of titles) {
      await st.step(`"${title}"`, async (sst) => {
        await assertSnapshot(sst, extractFromCodeTitle(title));
      });
    }
  });

  await t.step("inaccurate titles", async (st) => {
    const nonTitles = [
      "  code: foo. ", // Invalid: Trailing dot without extension is not a valid code block format
      // Returning `null` is expected as this format is invalid
      "any:code: foo ", // Invalid: Must start with exactly "code:" prefix
      " I'm not code block ", // Invalid: Not a code block format at all
    ];
    for (const title of nonTitles) {
      await st.step(`"${title}"`, async () => {
        await assertEquals(null, extractFromCodeTitle(title));
      });
    }
  });
});
