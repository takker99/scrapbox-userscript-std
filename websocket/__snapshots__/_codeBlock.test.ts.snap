export const snapshot = {};

snapshot[`extractFromCodeTitle() > accurate titles > "code:foo.extA(extB)" 1`] = `
{
  filename: "foo.extA",
  indent: 0,
  lang: "extB",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > " code:foo.extA(extB)" 1`] = `
{
  filename: "foo.extA",
  indent: 1,
  lang: "extB",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > "  code: foo.extA (extB)" 1`] = `
{
  filename: "foo.extA",
  indent: 2,
  lang: "extB",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > "  code: foo (extB) " 1`] = `
{
  filename: "foo",
  indent: 2,
  lang: "extB",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > "  code: foo.extA " 1`] = `
{
  filename: "foo.extA",
  indent: 2,
  lang: "extA",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > "  code: foo " 1`] = `
{
  filename: "foo",
  indent: 2,
  lang: "foo",
}
`;

snapshot[`extractFromCodeTitle() > accurate titles > "  code: .foo " 1`] = `
{
  filename: ".foo",
  indent: 2,
  lang: ".foo",
}
`;
